### PR TITLE
Tool packages

### PR DESCRIPTION
**Pull request by tcdent**
# Tool Packages

This PR introduces a new way of packaging tools in AgentStack. Tools are now wrapped in a framework-specific layer to allow for our implementations to remain completely generic. Some tools have been reimplemented to work with all frameworks instead of just importing from crewai-tools.

## Core Changes
- Tools now get wrapped in a framework-specific layer
- Reimplemented tools to work with all frameworks instead of just importing from crewai-tools
- Removed `tools_bundled` from ToolConfig as all tools are now bundled for namespacing
- `agentstack.tools` is not part of the public API and is to be used inside a user project to access tools

## Directory Search Implementation
As part of this effort, the directory search tool has been reimplemented as framework-agnostic functions:

- Added two core functions:
  - `search_directory(directory: str, query: str)`: Search any directory
  - `search_fixed_directory(query: str)`: Search preconfigured directory
- Removed dependency on CrewAI's BaseTool
- Uses embedchain.loaders.DirectoryLoader directly
- Framework-specific wrapping handled by frameworks.crewai.get_tool_callables

### Environment Variables
- `DIRECTORY_SEARCH_PATH`: Optional. Path to directory for fixed directory search.

## Composio Tool Implementation
The composio tool has been reimplemented as framework-agnostic functions:

- Added core functions:
  - `execute_action`: Execute specified Composio actions with parameters
  - `get_action_schema`: Retrieve schema for Composio actions
  - `find_actions_by_use_case`: Find actions based on use case
  - `find_actions_by_tags`: Locate actions by specified tags
- Removed dependency on CrewAI's BaseTool
- Uses composio library directly for core functionality
- Framework-specific wrapping handled automatically by agentstack.tools['composio']
- Updated config.json with proper dependencies and CTA field

### Environment Variables
- `COMPOSIO_API_KEY`: Required. API key for Composio service access.

## Stripe Tool Implementation
The Stripe tool has been reimplemented as framework-agnostic functions:
- Removed framework-specific dependencies
- Implemented core functions for:
  - Payment links (create/read)
  - Products (create/update)
  - Prices (create/update)
- Uses stripe library directly for core functionality
- Framework-specific wrapping handled automatically

### Environment Variables
- `STRIPE_SECRET_KEY`: Required. API key for Stripe service access.

Link to Devin run: https://app.devin.ai/sessions/24be273939424e7eacea38661326111e

### CODE REVIEW

**Comment by bboynton97** [comment]
lets save this for 0.3 because it will require the project upgrade script. we can merge it now though and create a 0.2 branch for bug fixes and small features

**Comment by bboynton97** [comment]
one challenge with this implementation is that by not copying the tool into the user's project, we remove the ability for them to modify the tool to best fit their use-case. So for Composio for example, there are a number of ways in which the tool can be adjusted for a specific implementation.

additionally, bundling all tools isn't the best idea. if a tool provider exports 20 tools and the developer only wants one of them, it consumes additional context to give the LLM all 20 tools, as well as creating the opportunity for the LLM to pick the wrong tool.

**Comment by tcdent** [comment]
> users can still copy the tool source into their project tools dir to modify the implementation

We can add an `-e --editable` flag to copy the source to the user's project. 

I've been thinking about how we handle custom tools in templates. Still don't have a bright idea, but it would be cool to still be able to package up a project with custom tools and share the template. 

>> additionally, bundling all tools isn't the best idea. if a tool provider exports 20 tools and the developer only wants one of them, it consumes additional context to give the LLM all 20 tools, as well as creating the opportunity for the LLM to pick the wrong tool.

> will think about this. I wanted to incorporate namespacing so it was clear which tool the function was coming from. previously, installing a tool added all of the functions inside the tool package.

allowing the user to select which functions they're installing, in addition to the tool, using the new interactive `agentstack tools add` interface would be easy enough. 

`*agentstack.tools['tool_name']` for all tools
`agentstack.tools['tool_name']['function_name'], ...` for selected tools
`tools.function_name` for tools inside the user's project

just make the branching inside the framework ast editability a bit more complex, but very doable. 

moving the installed tools from `agentstack.json` to `conf/tools.yaml` would let us track this better

File: agentstack/_tools/directory_search/config.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "directory_search",
+  "category": "computer-control",
+  "tools": ["search_directory", "search_fixed_directory"],
+  "description": "Search through files in a directory using embedchain's DirectoryLoader",
+  "env": {
+    "DIRECTORY_SEARCH_TOOL_PATH": "Optional. Path to directory for fixed directory search."
```
Comment (bboynton97) [inline agentstack/_tools/directory_search/config.json:7]: `null`

File: agentstack/_tools/firecrawl/config.json [hunk 2]
```diff
@@ -2,10 +2,12 @@
   "name": "firecrawl",
   "url": "https://www.firecrawl.dev/",
   "category": "browsing",
-  "packages": ["firecrawl-py"],
   "env": {
     "FIRECRAWL_API_KEY": "..."
```
Comment (bboynton97) [inline agentstack/_tools/firecrawl/config.json:6]: `none`

File: agentstack/_tools/neon/config.json [hunk 3]
```diff
@@ -2,10 +2,13 @@
     "name": "neon",
     "category": "database",
     "url": "https://github.com/neondatabase/neon",
-    "packages": ["neon-api", "psycopg2-binary"],
     "env": {
         "NEON_API_KEY": "..."
```
Comment (bboynton97) [inline agentstack/_tools/neon/config.json:6]: `null`

File: agentstack/_tools/~README.md [hunk 4]
```diff
@@ -1,5 +1,30 @@
-Tool Configuration Files
-========================
+
+
+Tool Implementations
```
Comment (bboynton97) [inline agentstack/_tools/~README.md:3]: update to new structure

File: agentstack/frameworks/crewai.py [hunk 5]
```diff
@@ -294,3 +324,20 @@ def remove_tool(tool: ToolConfig, agent_name: str):
...
+    tool_funcs = []
+    tool_config = ToolConfig.from_tool_name(tool_name)
+    for tool_func_name in tool_config.tools:
+        tool_func = getattr(tool_config.module, tool_func_name)
+
+        assert callable(tool_func), f"Tool function {tool_func_name} is not callable."
+        assert tool_func.__doc__, f"Tool function {tool_func_name} is missing a docstring."
+
+        # apply the CrewAI tool decorator to the tool function
+        tool_funcs.append(_crewai_tool_decorator(tool_func))
```
Comment (bboynton97) [inline agentstack/frameworks/crewai.py:342]: 🧼

File: agentstack/generation/files.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 from typing import Optional, Union
+import string
```
Comment (bboynton97) [inline agentstack/generation/files.py:2]: merge this back

File: agentstack/generation/tool_generation.py [hunk 7]
```diff
@@ -132,21 +54,11 @@ def remove_tool(tool_name: str, agents: Optional[list[str]] = []):
         print(term_color(f'Tool {tool_name} is not installed', 'red'))
         sys.exit(1)
 
+    # TODO ensure other agents are not using the tool
```
Comment (bboynton97) [inline agentstack/generation/tool_generation.py:57]: lets make an issue for this

File: agentstack/generation/tool_generation.py [hunk 8]
```diff
@@ -132,21 +54,11 @@ def remove_tool(tool_name: str, agents: Optional[list[str]] = []):
...
-    except FileNotFoundError:
-        print(f'"src/tools/{tool.module_name}.py" not found')
-
-    try:  # Edit the user's project tool init file to exclude the tool
-        with ToolsInitFile(conf.PATH / TOOLS_INIT_FILENAME) as tools_init:
-            tools_init.remove_import_for_tool(tool, agentstack_config.framework)
-    except ValidationError as e:
-        print(term_color(f"Error removing tool:\n{e}", 'red'))
+    if tool.dependencies:
+        # TODO split on "==", ">=", etc. and only remove by package name
```
Comment (bboynton97) [inline agentstack/generation/tool_generation.py:60]: ^

**Review by braelyn-ai — APPROVED** [review]
tl;dr
lgtm

**Comment by potpiedev** [comment]
Starting Review, should be available shortly
